포커게임 앱 Step4 미션 제출합니다.

ehgud0670 opened the pull request:
> 고민한 점

* 테스트할 때 해당 객체의 프로퍼티에 접근할 때 get 방식이 아닌 클로져 형태로 하려는 방식을 고민하고 
  코드로  구현하였습니다. 
* 플레이어들이 카드를 받을 때 클로저를 통해 받도록 하려는 방식을 고민하고 
  코드로 구현하였습니다. 

> 아쉬운 점
* PokerGame 객체의 startNewRound()에 대한 테스트코드를 카드들의 개수를 비교하는 식으로
간접적으로 확인하는 방법으로 작성했지, 정통으로 해당 메서드가 완벽한지 테스트하는 코드는
작성하지 못하였습니다. startNewRound() 안에 shuffle()메서드가 호출되는데 이 shuffle()에 대해 의존성을
없애야지만 가능할 것 같습니다. 

> 애매한 점 
* PokerGame 에서 셔플이 더 나아지도록 하기위해(?) generator 프로퍼티를 선언하고 
shuffle() 함수에서 사용되는데 shuffle()이 프로퍼티인 generator를 사용하므로 순수함수가 아닌 것이 
마음에 걸립니다.

ehgud0670 committed:
Refactor : 덱 객체의 카드배열 프로퍼티가 외부에 안보이도록 private 추가

ehgud0670 committed:
Refactor : 랜덤 메서드인 suffle에 대한 테스트 코드 기능 수정
          * 고정된 랜덤 넘버 제너레이터인 ANSI_C_RandomNumberGenerator를 이용해
	    랜덤 메서드 shuffle에 대한 테스트가 100 % 정확하도록 함

ehgud0670 committed:
Refactor : suffle 테스트할때 클로져로 카드들을 가져와 카드들끼리 비교되도록 기능 수정
	* card 객체의 프로퍼티는 모두 private 하므로 클로져로 가져와도 된다고 생각해서 기능 수정함

ehgud0670 committed:
Refactor : 덱 객체에서 cards을 초기화하는 메소드를 삭제하고 싱글톤 cards 배열 추가
	이유 : reset할때마다 초기화하는 메소드를 부르는 것은 비효율적이다.
		따라서 대신 싱글톤 cards 배열을 프로퍼티로 추가하고 이 배열을 할당하도록 함.

ehgud0670 committed:
Refactor : 카드의 타입을 class에서 struct로 변경
	이유 : 1. 카드는 모델 객체라 되도록이면 변화를 피하기 위해 struct로 변경
	       2. 카드는 필수적으로 카드덱의 프로퍼티이므로 이후에
		혹시모를 강한 참조 cycle을 피하기 위해 struct로 변경함.

ehgud0670 committed:
Style : ':' 앞의 Spacing 한칸씩 지워서 style 통일되게 함

ehgud0670 committed:
Refactor : 싱글톤 객체가 되도록 static 타입으로 변경

ehgud0670 committed:
Feature : 게임 플레이어 객체 추가
         * 프로퍼티로 옵셔널? 타입인 카드배열 객체 추가

         게임 플레이어가 class인 이유
	    : 이후에 나타날 게임 판에서 딜러를 Player 타입으로 두고 let으로 선언해 constant로 두고 싶다.
              따라서 이후 딜러의 속성들을 변경하려면 딜러의 타입이 struct가 아닌 class 이어야 하므로
              Player타입을 class로 두었다.

         카드배열의 타입이 옵셔널?인 이유
            : 1. 플레이어의 카드들은 이후에 나타날 게임 판의 카드덱에 의존적이다.
            : 2. 이후 미션에서 앱을 shake()할때 플레어이들에게 카드들을 다시 나눠줘야 하고
            이때 플레이어의 카드 배열이 nil이 된 이후 다시 세팅될 가능성이 많으므로
            안전성을 위해 옵셔녈?로 두었다.

ehgud0670 committed:
Feature : Player가 탐색되도록 Searchable를 채택하게 코드 작성
	* map을 사용함
	* deck struct에서도 map을 사용하도록 코드 수정

ehgud0670 committed:
Feature : Player가 Playerable을 준수하도록 기능 구현
	* 클로져를 통해 카드를 받도록 메소드 구현

ehgud0670 committed:
Style : 카드를 탐색한다는 의미를 부여하기 위해 메소드 이름 변경

ehgud0670 committed:
Refactor : 카드덱의 타입을 struct에서 class로 변경
	* 이유 : 이후의 게임 객체가 프로퍼티로 카드덱을 가질텐데
	카드덱을 let으로 선언하고 싶다.
	따라서 카드덱의 프로퍼티는 변경가능하게 카드덱의 타입을 class로 변경

ehgud0670 committed:
Feature : 포커게임 클래스 프로퍼티 선언 및 초기화
	  * 카드덱 프로퍼티 let으로 초기화
	  * 딜러 프로퍼티를 let, Player객체로 초기화
	  * 프로퍼티인 플레이어들을 Player 배열 옵셔널! 타입으로 선언
	  => 게임 진행할때 반드시 필요한 객체들이므로 !으로 선언
	  => 바로 접근가능하도록 !으로 선언

	  * 포커게임을 클래스로 구현한 이유
	   => 이후에 Controller로 올라갈건데 let으로 두고
	  포커게임의 프로퍼티만 변경가능하도록 하고 싶기 때문.

ehgud0670 committed:
Feature : PokerGame 클래스 이니셜라이저 추가
	* 입력값 playerNumber가 1이상 4이하가 아니면 에러를 던지도록 하고
	* PokerGame 객체가 생성되지 않도록 init?으로 작성함

ehgud0670 committed:
Refactor : 포커게임 클래스의 프로퍼티의 접근범위를 모두 private으로 변경

ehgud0670 committed:
Feature : * 게임 스터드 방식을 나타내는 enum 선언
	  * gameStut을 게임 클래스의 프로퍼티로 선언

ehgud0670 committed:
Refactor : 포커게임 클래스의 이니셜라이져 파라미터 변경
	  * GameStut 타입에서 Int로 변경
	  * 적절한 gameStut의 rawValue가 아닌 경우 에러 던지도록
	    기능 구현

	* 이유 : 게임과 관련된 클래스,구조체, enum들을
	  viewController로 부터 캡슐화하고 싶어서

ehgud0670 committed:
Feature : 게임 객체의 이니셜라이저에 대한 테스트 기능 구현
         * 매개변수 gameStutNumber 에 대한 테스트
         * 매개변수 playersNumber 에 대한 테스트

ehgud0670 committed:
Feature : 카드가 나눠줄정도로 충분한지 아닌지 나타내는 기능 구현
	* 외부에서 사용가능하도록 interal 접근 제한자로 설정

ehgud0670 committed:
Style : stut 연산프로퍼티 위치 변경

ehgud0670 committed:
Feature : 포커게임이 카드덱을 셔플하도록 기능 구현
	* 외부에서 안보이도록 private 으로 선언
	* 회마다 다른 seed의 generator를 사용하도록
	프로퍼티인 generator 선언

ehgud0670 committed:
Refactor : 제너레이터 외부에서 안보이도록 private으로 선언

ehgud0670 committed:
Refactor : 게임 라운드 시작하면 카드덱이 셔플 되도록 기능 구현
	 * 외부에서 호출되도록 메소드 접근 제한자 internal로 적용:

ehgud0670 committed:
Feature : 게임 라운드 시작하면 딜러와 플레이어들에게 카드 나눠주도록 기능 구현
	* try문이 많은데 deck의 removeOne()으로 인한 것이고
	 가장 상위에서 에러를 잡도록 모두 try로 에러를 넘겨주도록 기능 구현
	* 가독성을 위해 여러 메서드로 나눠서 기능 구현

ehgud0670 committed:
Feature : 게임 객체의 hasEnoughCards에 대한 테스트 기능 구현

ehgud0670 committed:
Refactor : 게임 라운드가 시작할 때마다 플레이어들의 카드들 init되도록 기능 수정
	* 옵셔널 타입인 플레이어의 cards가 초기화되도록 플레이어의 메소드 기능 추가

ehgud0670 committed:
Feature : 게임 객체의 startNewRound 메소드의 테스트 코드 구현
	* 게임 객체의 프로퍼티들을 탐색하도록 메소드들 구현
		* 파라미터로 클로저 사용

ehgud0670 committed:
Modified : step4 의 내용에 맞게 README.md 파일 작성

ehgud0670 committed:
Refactor : Deck이 클래스 타입인 것에 맞게 테스트 코드 수정

ehgud0670 committed:
Refactor : 게임 객체의 프로퍼티인 덱 객체를 상수로 변경

godrm commented on CardGameApp/CardGameApp/Domain/Card.swift:
```diff
@@ -62,7 +62,7 @@ extension Card.Suit : CustomStringConvertible {
+extension Card.Number: CustomStringConvertible {
     var description: String {
         switch self {
         case .ace:
```
프로토콜을 채택해서 선언 부분에 함께 구현하는 것과, 지금처럼 extension으로 분리하는 것과 어떤 차이가 있을까요?
어떤 기준에서 작성하는지 궁금합니다. 
저는 사용하는 곳보다는 항상 위에 쓰는 편입니다. 그래야 어디에 있는지 찾기가 수월하더라구요.

ehgud0670 replied:
extension은 확장의 의미로 본래 목적의 클래스를 수정하지 않고 확장하는 Open-Closed 원칙에 부합하여 사용하였습니다. 
첫 미션에서 카드의 넘버나 Suit를 문자열로 출력할 이유가 없었지만 이후의 요구사항에서 필요했으므로
본래의 코드를 수정하지않고 확장하기 위해 extension을 사용하였습니다. 

그리고 사용하는 곳 보다 위에 쓰기 위해 코드를 수정하겠습니다. 감사합니다.

godrm commented on CardGameApp/CardGameApp/Domain/Deck.swift:
```diff
@@ -45,17 +45,27 @@ struct Deck {
         return cards.remove(at: indexFirstCard)
```
이렇게 Error를 throw 하는 것과 Card? 옵셔널 처리하는 것과 어떤 차이가 있을까요?
왜 에러를 던지는 것으로 구현했는지 궁금합니다.

ehgud0670 replied:
리턴할 값을 옵셔널로 주는 것은 런타임 에러를 주지 않는다는 점에서 좋은 접근이지만, 
Error를 throw를 하는 것은 런타임 에러를 갖지 않으면서도 어느 부분에서 잘못되었는지 로그를 볼 수 있는 형식이기 때문에 
더 좋은 판단이라 생각해 Error를 throw를 했습니다. 그리고 이렇게 error를 계속 throw해서 최상위 객체까지 끌고 가면 전체적으로 어디서 에러가 났는지 파악이 되기 때문에 좋은 방법이라고 들어서 이런식으로 코드를 작성하였습니다.  

`그리고 이렇게 error를 계속 throw해서 최상위 객체까지 끌고 가면 전체적으로 어디서 에러가 났는지 파악이 되기 때문에`
=> 주워들은 거라 100% 확신은 못하겠습니만, 에러가 잘 파악되긴 할 것 같습니다. 맞는 말인 걸까요

ehgud0670 replied:
`이렇게 error를 계속 throw해서 최상위 객체까지 끌고 가면 전체적으로 어디서 에러가 났는지 파악이 되기 때문에 좋은 방법이라고 들어서`
=> 이거 맞는 말일까요?? 궁금합니다 ㅎ 🤓

godrm commented on CardGameApp/CardGameApp/Domain/Deck.swift:
```diff
@@ -45,17 +45,27 @@ struct Deck {
+
+extension Deck: Searchable {
+    func searchCard(handler: (Card) -> ()) {
+        _ = cards.map { handler($0)}
```
map()을 사용하는 것과 forEach()를 사용하는 것은 어떤 차이가 있을까요?
왜 map()을 사용했는지 궁금합니다.

ehgud0670 replied:
사실 저번 cs때 map을 배워서 별 생각없이 map을 사용했습니다. 
찾아보니 map은 클로져로 매핑한 결과의 배열을 반환할때 사용하고 
forEach는 for-in 루프와 동일한 순서로 시퀀스의 각 요소에 대해 주어진 클로저를 호출하고 싶을때 사용한다고 하네요(출처 : https://stackoverflow.com/questions/45449977/array-map-vs-foreach-in-swift)

=> 따라서 지금 코드에서는 forEach를 사용하는 것이 적절한 것 같아요. 고치겠습니다! 👍

godrm commented on CardGameApp/CardGameApp/Domain/Player.swift:
```diff
@@ -0,0 +1,36 @@
+
+class Player: Playerable , Searchable {
+    
+    private var cards : [Card]?
```
[Card]? 옵셔널로 지정한 이유가 궁금합니다.
아래 보면 init() 함수에서 항상 빈 배열을 초기화하는 것 같네요.
널(nil), 빈배열, 값이 있는 상태 이렇게 3개를 구분한 이유가 궁금합니다.

ehgud0670 replied:
게임이 시작할때 플레이어들의 카드들을 초기화하고 싶어서 이렇게 구현했습니다. 
그리고 이후에 앱을 흔들면 다시 나눠주게 되는데 원래 있던 카드들을 init()함수를 호출해서 비우고 다시 받게끔 구현하고 싶었습니다. 답변 부탁드립니다!

godrm replied:
init은 인스턴스를 생성하는 초기화를 의미합니다. 다시 내용을 복구하는게 아니거든요. 
그래서 내부 변수가 항상 초기화하는 곳으로 사용하는 거죠. 
이전에 init을 한 객체를 init을 다시 호출하는 경우는 없습니다.
제가 궁금했던 것은 nil 상태가 정말 필요하냐는 겁니다. 늘 초기화가 된다면 옵셔널일 필요 없을 수도 있으니까요.

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+        func forEach(handler: () throws  -> ()) throws {
+            for _ in 0 ..< self.rawValue {
+                try handler()
+            }
```
위에 forEach 와 아래 forEach는 어떻게 다른가요? 어떻게 구분해서 호출해야 하죠?
선언 부분만 봐서는 구분이 전혀 안되네요. 어떤 의도가 있는거죠?

ehgud0670 replied:
아래의 forEach문은 Deck의 removeOne()이 에러를 던지기 때문에 거기에 맞게 작성한 메서드입니다.  
클로저가 에러를 던지면 아래 forEach 메서드를 쓰고 클로저가 에러를 던지지 않으면 위의 forEach 메서드를 사용하면 되겠다고 생각하고 
두 메서드를 작성하였습니다. 
아래의 forEach 메서드 명을 좀 바꿔서 구분을 하면 될까요??

godrm replied:
같은 이름을 두 개 만들필요는 없죠. 더구나 에러처리 하고 안하고 그런건 구분할 필요없이 하나만 만들어도 될 것 같아서요. 전혀 다른 기능도 아니고 동일한 기능은 하나만 있으면 되겠죠.

ehgud0670 replied:
엇 근데 handler가 될 클로져가 에러를 던지면 아래 forEach 메서드의 경우이고, 클로져가 에러를 던지지 않으면 위 forEach 메서드 경우인데 통합할 수 있을까요? 위의 forEach 메서드를 없애고 
아래 forEach문을 호출해서 에러를 던지지 않는 클로저라도 try 문을 쓰게끔 할까요?

godrm replied:
클로저로 핸들러를 넘기는 경우는 성공할 때 클로저와 실패할 때 클로저를 모두 넘기는 경우도 있습니다. 참고하세요. 
꼭 에러를 던져야 하는지도 고민해봐야죠. 클로저에서 에러가 난다는 게 어떤 의미일까요?
매개변수로 처리하는 경우도 있습니다. Error나 Result<> 타입을 활용하기도 합니다.

ehgud0670 replied:
실패할 때를  대비하여 Error? 를 매개변수로 활용한 것을 DNS Server 할때 본 적이있습니다. 
**Result<>타입은 몰라서 추후에 공부해보겠습니다.**

ehgud0670 replied:
클로저에서 에러가 난다는 것은 런타임시 일어난 에러를 잡기 어려운 상황인 것같은데 잘 모르겠습니다.ㅠ

ehgud0670 replied:
매개변수인 클로저에서 에러가 난다는 것은 내가 소유하지않은 함수에서 에러가 난다는 것을 뜻합니다. 
또 비동기적인 상황에서 클로저를 매개변수로 사용하는 경우가 많이 있는데 이러한 경우 비 동기적으로 에러가 난다는 것을 뜻합니다.

godrm replied:
어.. 그렇군요. 그렇게 클로저 내부에서 에러가 발생할 정도라면 클로저를 사용하면 안되는 경우일지도 모릅니다. 모든걸 클로저로 해야한다는 의미는 아니에요. 
그리고 클로저에서 에러가 발생하면 클로저 내부에서 해결해야 할겁니다. 왜냐하면 클로저를 실행하는 곳에서는 어떤 예외가 발생할 지 모르니까요

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+        case five = 5, seven = 7
+        
+        func isValid(num: Int) -> Bool {
+            return self.rawValue == num
```
이 비교가 왜 필요한가요? init 할 때 검증할 수 있지 않나요?

ehgud0670 replied:
생각해보도록 하겠습니다.

ehgud0670 replied:
아 넵. 그 방법이 더 나은 것같습니다. 지우겠습니다.

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+    private let dealer = Player()
+    private var players: [Player]!
+    
+    private enum GameStut: Int {
```
이정도로 구현코드가 길어지면 내부 nested enum 이면 코드 읽기 어려운데 분리는 가능한가요? 전혀 불가능한가요?

ehgud0670 replied:
분리해보도록 하겠습니다.

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+        guard playersNumber >= minPlayersNumber,
+            playersNumber <= maxPlayersNumber else {
+                throw PokerGameError.invalidPlayersNumber
+        }
```
비교문으로 런타임에 비교하지 마세요. 값의 범위가 바뀌면 enum 타입만 바꾸면 되는 구조가 더 좋습니다. 
매번 비교문을 수정하는 것은 좋지 않습니다. 매번 테스트 코스도 달라질테니까요.

ehgud0670 replied:
* 넵. 저도 코드를 작성하면서 분명히 gameStut의 enum의 개수가 늘어나거나 playerNumber의 범위가 늘어나면 계속 변경해야되서 번거로울 것 같다는 생각을 했습니다. 고치겠습니다.

* 다만, swift도 java와 접근 제한자의 범위가 똑같을 것 같다는 착각으로 패키지(폴더)가 다르면 internal(default) 제한자로 인해 **상위의 폴더**에서 사용을 못할 것 같다는 생각에 (**상위의 폴더**에서 사용을 안했으면 했습니다.) 이후에 viewcontroller에서 Pokergame의 매개변수로 Int값(stutNum, playersNum)을 주려고 했습니다.  

근데 internal 제한자 모듈 안에서 모두 사용가능하므로 상위의 폴더에서 GameStut이나 Players를 사용할 수 있어 해당 논리는 소용이 없었습니다ㅋㅋ 😅

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+        for _ in 0 ..< num {
+            players.append(Player())
+        }
+        return players
```
[Player]를 담당하는 Players 라는 타입을 만들어보세요.

ehgud0670 replied:
고쳐보도록 하겠습니다 👍

godrm commented on CardGameApp/CardGameApp/Domain/PokerGame.swift:
```diff
@@ -0,0 +1,141 @@
+            try players[index].receiveCard{
+                try deck.removeOne()
+            }
+        }
```
위에 Players 라는 타입을 만들면 여기 반복문이 Players 내부로 옮겨지고 단순해질 겁니다.

ehgud0670 replied:
넵 👍

godrm commented on CardGameApp/CardGameAppTests/DeckTests.swift:
```diff
@@ -58,25 +61,28 @@ class CardGameAppTests: XCTestCase {
-        try! otherDeck!.removeOne()
-        XCTAssertEqual(otherDeck?.count, totalCardCounts - 2 )
+        try! deck.removeOne()
+        XCTAssertEqual(deck.count, totalCardCounts - 2 )
```
테스트 함수 이름도 testScenario() 보다 더 구체적으로 어떤 동작 시나리오인지 명시하는 게 좋습니다.
함수 이름만 보고 무엇을 실패했는지 알 수 있어야 합니다.

ehgud0670 replied:
넵 명심하겠습니다, 고치겠습니다.

godrm commented on CardGameApp/CardGameAppTests/PokerGameTests.swift:
```diff
@@ -0,0 +1,107 @@
+        let correctNumberSeven = 7
+        game = try? PokerGame(gameStutNumber: correctNumberSeven,
+                              playersNumber: correctPlayersNum)
+        XCTAssertNotNil(game)
```
값이 다르면 다른 테스트 함수로 분리하세요. 
3일 때 테스트 케이스가 실패한건지, 5일때 실패한건지 알 수 없으니까요.

ehgud0670 replied:
네넵 분리해보도록 하겠습니다.

godrm reviewed (commented):
전체적인 구조는 나쁘지 않습니다. 
세부적인 개선을 위해서 다음 피드백을 확인해보세요.

ehgud0670 committed:
Refactor : PokerGame init 코드 변경 및 nested enum 분리
          * PlayersNum enum 생성
          * GameStut enum과 PlayersNum enum 을 생성자의 파라미터값으로 변경함
	  * GameStur enum 의 isValid() 메서드 삭제

	* 생성자의 파라미터 타입을 바꾼 이유
	: Int값으로 매개변수를 받아서 최대값과 최솟값을 변경하는 것은
	이후에 코드를 변경할 때마다 테스트 코드도 재작성해야 하므로 번거롭기 때문

	* GameStut enum의 isValid()을 없앤 이유:
	단순히 init() 하는 것이 더 깔끔하고 더 나은 방법이기 때문이다.
	GameStut(.five) 나 GameStut(.seven) 이런 방식이 더 깔끔하고
	개발자가 enum의 case들만 선택하면 되기 때문에 오류가 날 일이 없기 때문이다.

ehgud0670 committed:
Refactor : 클로져에서 error를 throw하는 대신 error를 반환값으로 보내는 것으로 변경

           * GameStut의 forEach 메소드에서 클로저의 반환 타입을 Error?로 변경
           => 이유 : 1. 이전의 2개의 forEach문을 통합하기 위해서
                     2. 결국에는 에러를 전달하고 싶은데 throw 방식보다는 반환하는 것이
			나을 것이라고 판단했기 때문

	   * playerable의 receiveCard의 반환값을 2개(card : Card?, error : Error?)로 변경함

ehgud0670 commented:
아직 모든 피드백을 수정하지 않았습니다만,
저의 error를 처리하는 방식이 올바른건지 궁금해서 먼저 푸시했습니다. 
**저는 receiveCard에서 에러를 표시하기 위해 기존의 card 반환값에서 추가로 error를 반환값을 추가했는데 
올바른 방법일까요??**( 근데 반환값이 2개가 되는데 신기하네요 ㅋㅋ)
결국엔 ResultType을 사용할건데 이 방법이 올바른지 궁금합니다.


(매개변수로 클로저를 사용할때 Standard library의 클로저도 많이 사용해서 클로져가 일반적이어야 하기때문에 안될것같기도 합니다. This is works well if you own that code, but if you’re using one of the standard library functions then not so much.  => <https://andybargh.com/error-handling-in-swift/#Error_Propagation_With_Asynchronous_Closures>)


==> sender() 두번 호출하면 두번 카드가 없어지네요.. 그냥 빨리 result<>로 작성하겠습니다.

ehgud0670 committed:
Deleted : 상황을 설명할 수 없는 testScenario() 테스트 코드 삭제

ehgud0670 committed:
Refactor : 반환타입 Result<Card,Error>로 변경

ehgud0670 committed:
Refactor : map에서 forEach로 변경

ehgud0670 committed:
Refactor : PokerGame을 확장해서 nested enum을 분리 및 Card의 extension 위치 변경

ehgud0670 committed:
Refactor : enum을 확장함에 따라 테스트 코드 수정

ehgud0670 commented:
피드백대로 코드를 수정하였습니다. 

* PokerGame 초기화 코드 변경
* 클로저에서 반환타입 Result<>로 변경
* 배열의 메소드 map => forEach 로 변경
* PokerGame의 neste enum들 확장해서 분리함
* Card의 extension 코드 위치를 변경 
* 테스트 코드 재작성

ehgud0670 commented:
Result타입을 사용한 이유는 이번엔 한번 사용해보고 싶었습니다..
이후에 클로저 내부에서 에러를 잡거나, 클로저 내부에서 에러가 안나도록 removeOne의 메소드를 변경해보겠습니닷

godrm commented on CardGameApp/CardGameApp/Domain/Player.swift:
```diff
@@ -9,28 +9,32 @@
+            cards?.append(card)
+        case .failure(let error):
+            throw error
+        }
```
이 부분은 좀 어색하네요. 
closure가 매개변수로 넘어온다고 해서, 그걸 호출해서 결과를 처리하는 방식보다는
Player 입장에서는 그냥 Card만 넘어오면 되는거 아닐까요?
sender() 라는 클로저는 여기서 실행할 필요가 없습니다. 어떤게 실행될지 전혀 알수가 없으니까 예외처리도 이상해집니다. 
새로운 에러 코드가 추가되도 알 방법이 없으니까요.

godrm commented:
마무리하고 다음 단계를 진행해보세요.
